app/models/bookmark.rb

```diff
@@ -156,11 +165,11 @@ def self.cleanup!
 #
 # Indexes
 #
-#  index_bookmarks_on_post_id              (post_id)
-#  index_bookmarks_on_reminder_at          (reminder_at)
-#  index_bookmarks_on_reminder_set_at      (reminder_set_at)
-#  index_bookmarks_on_reminder_type        (reminder_type)
-#  index_bookmarks_on_topic_id             (topic_id)
-#  index_bookmarks_on_user_id              (user_id)
-#  index_bookmarks_on_user_id_and_post_id  (user_id,post_id) UNIQUE
+#  index_bookmarks_on_post_id                           (post_id)
+#  index_bookmarks_on_reminder_at                       (reminder_at)
+#  index_bookmarks_on_reminder_set_at                   (reminder_set_at)
+#  index_bookmarks_on_reminder_type                     (reminder_type)
+#  index_bookmarks_on_topic_id                          (topic_id)
+#  index_bookmarks_on_user_id                           (user_id)
+#  index_bookmarks_on_user_id_and_post_id_and_topic_id  (user_id,post_id,topic_id) UNIQUE
```

Comment:
Am I right to say that if `post_id` is present, `topic_id` will be nil and vice versa? If that is the case, shouldn't we maintain two index of `user_id, topic_id` and `user_id, post_id` instead?

Reply:
we would need to make them filtered for them to be unique , but yeah that can work.

one index for user_id,topic_id where post_id is null ... unique 
one index for user_id,post_id where topic_id is null ... unique 

I guess one big question is if we are relying on de-normalization here for any of the queries eg bookmark post X in topic Y, then move post X to topic M...

Reply:
No, topic ID will always be present, only post_id is nullable. So this index should work for:

`user_id: 1, topic_id: 99, post_id: 123`

and

`user_id: 1, topic_id: 99, post_id: 124`

and

`user_id: 1, topic_id: 99, post_id: NULL`

Right?

Reply:
I think the index is fine for that constraint, but I guess, what happens when you move a bookmarked post from one topic to another, does it vanish or is there code somewhere that corrects this?